File: src/sprites.js
/***********************************************	DRONE 		*******************/
var INITIAL_DRONE_SPAWN_AMOUNT = 20;
var droneSpawnAmount = INITIAL_DRONE_SPAWN_AMOUNT;
var droneNextSpawnTime; //time till next spawn (random b/t 3-5 seconds)
var drone = {
	radius: 9,
	dead: false, //to prevent a weird afterdeath collision bug
	color: "Cyan",

	xPos: 0,
	yPos: 0,
	
	movementSpeed: 6,
	xVel: 0,
	yVel: 0,
	pts: 15, //how many pts the player gets for killing this drone

	moveToPos: function(newX, newY) { //decomposes velocity into appropriate x and y based off drones sum total movement speed and sets target x and y
		var dx = newX - this.xPos;
		var dy = newY - this.yPos;
		var angle = Math.atan2(dy, dx)

		this.xVel = this.movementSpeed * Math.cos(angle);
		this.yVel = this.movementSpeed * Math.sin(angle);
	},

	updatePosition: function() {
		this.moveToPos(player.xPos, player.yPos); //target player

		//Move the player and keep it inside screen boundaries
		this.xPos = Math.max(0 + this.radius, Math.min(this.xPos + this.xVel, canvas.width - this.radius));
		this.yPos = Math.max(0 + this.radius, Math.min(this.yPos + this.yVel, canvas.height - this.radius));
	},

	draw: function() {
		//draw drone as circle
		drawingSurface.beginPath();
		drawingSurface.arc(Math.floor(this.xPos), Math.floor(this.yPos), this.radius, 0, 2 * Math.PI, false);
		drawingSurface.fillStyle = this.color;
      	drawingSurface.fill();	
	},

	spawn: function(corner) { 
		var minX = 0;
		var maxX = 0;
		var minY = 0;
		var maxY = 0;
		//spawn in one of the four corners [0, 1, 2, 3]
		switch(corner) {
			case 0: //top left
				minX = 0;
				maxX = canvas.width / 4;
				minY = 0;
				maxY = canvas.height / 4;
				break;
			case 1: //top right
				minX = canvas.width * (3 / 4);
				maxX = canvas.width;
				minY = 0;
				maxY = canvas.height / 4;
				break;
			case 2: //bottom right
				minX = canvas.width * (3 / 4);
				maxX = canvas.width;
				minY = canvas.height * (3 / 4);
				maxY = canvas.height;
				break;
			case 3: //bottom left
				minX = 0;
				maxX = canvas.width / 4;
				minY = canvas.height * (3 / 4);
				maxY = canvas.height;
				break;
		}

		//offset by radius so that entire drone appears within boundaries (not half the radius outside)
		maxX -= this.radius;
		maxY -= this.radius;
		minX += this.radius;
		minY += this.radius;

		var randX = Math.floor(Math.random() * (maxX - minX) + minX); //randomX in range [drone.radius, maxX]
		var randY = Math.floor(Math.random() * (maxY - minY) + minY); 

		this.xPos = randX;
		this.yPos = randY;
		drones.push(this); //this is an array in the game.html
	},

	die: function() {
		this.dead = true;
		removeObjectFromArray(this, drones);

		var newExplosion = Object.create(explosion);
		newExplosion.xPos = this.xPos - this.radius;
		newExplosion.yPos = this.yPos - this.radius;
		//newExplosion.displaySize = this.radius;
		explosions.push(newExplosion);

		gameScore += this.pts;
	}
};
function spawnDrones() {
	drone_spawnMP3.volume = 0.3;
	drone_spawnMP3.play();

	var droneSpawnLoc = Math.floor(Math.random() * 4);
	while (droneSpawnLoc == player.inCorner()) { //don't spawn drones in the same corner as the player
		droneSpawnLoc = Math.floor(Math.random() * 4);
	}

	for (var i = 0; i < droneSpawnAmount; i++) {
		var newDrone = Object.create(drone);
		newDrone.spawn(droneSpawnLoc);	
	}

	//adjusted to gameSpeed
	var maxFreq = 5000 / gameSpeed;
	var minFreq = 3000 / gameSpeed;
	droneNextSpawnTime = Math.floor(Math.random() * (maxFreq - minFreq) + minFreq);
}

/***********************************************	EXPLOSION 		***************/
var explosion = {
	imgSrc: "burst.png",
	imgSize: 92,
	displaySize: 20, //pixels
	COLUMNS: 3, //change this to = the number of frames you have per row on your tile sheet
	numberOfFrames: 17, //18 - 1
	currentFrame: 0,
	sourceX: 0,
	sourceY: 0,

	xPos: 0,
	yPos: 0,

	updateAnimation: function() {
		//Find the frame's correct column and row on the tilesheet
		this.sourceX = Math.floor(this.currentFrame % this.COLUMNS) * this.imgSize;
		this.sourceY = Math.floor(this.currentFrame / this.COLUMNS) * this.imgSize;

		if (this.currentFrame < this.numberOfFrames) {
			this.currentFrame++;
		} else {
			removeObjectFromArray(this, explosions);
		}
	},

	draw: function() {
		//Draw the monster's current animation frame
		drawingSurface.drawImage (
			burstImage,
			this.sourceX, this.sourceY, this.imgSize, this.imgSize,
			this.xPos, this.yPos, this.displaySize, this.displaySize
		);
	}
}

/***********************************************	PLAYER 		*******************/
var player = {
	radius: 12,
	color: "white",
	xPos: 0,
	yPos: 0,
	
	movementSpeed: 8,
	moveLeft: false,
	moveTop: false,
	moveRight: false,
	moveDown: false,
	xVel: 0,
	yVel: 0,

	moveToPos: function(newX, newY) { //decomposes velocity into appropriate x and y based off drones sum total movement speed and sets target x and y
		var dx = newX - this.xPos;
		var dy = newY - this.yPos;
		var angle = Math.atan2(dy, dx)

		this.xVel = this.movementSpeed * Math.cos(angle);
		this.yVel = this.movementSpeed * Math.sin(angle);
	},

	updatePosition: function() {
		if (currentControlMethod == controlMethod.KBD) {
			//Up
			if (this.moveUp && !this.moveDown) {
				this.yVel = -this.movementSpeed;
			}
			//Down
			if (this.moveDown && !this.moveUp) {
				this.yVel = this.movementSpeed;
			}
			//Left
			if (this.moveLeft && !this.moveRight) {
				this.xVel = -this.movementSpeed;
			}
			//Right
			if (this.moveRight && !this.moveLeft) {
				this.xVel = this.movementSpeed;
			}

			//Set the player's velocity to zero if none of the keys are being pressed
			if (!this.moveUp && !this.moveDown) { //only set velocity to zero if mouse is not moving (otherwise we want the player to track the mouse)
				this.yVel = 0;
			}
			if (!this.moveLeft && !this.moveRight) {
				this.xVel = 0;
			}	
		} else if (currentControlMethod == controlMethod.MOUSE) { 
			//stop the player if he's within a certain distance of the mouse
			var errorFactor = 0.32;
			if (Math.abs(this.xPos - mouseX) < this.radius * errorFactor) { 
				this.xVel = 0;
			} 
			if (Math.abs(this.yPos - mouseY) < this.radius * errorFactor) {
				this.yVel = 0;
			}
		}

		//Move the player and keep it inside screen boundaries
		this.xPos = Math.max(0 + this.radius, Math.min(this.xPos + this.xVel, canvas.width - this.radius));
		this.yPos = Math.max(0 + this.radius, Math.min(this.yPos + this.yVel, canvas.height - this.radius));
	},

	draw: function() {
		//draw player as circle
		drawingSurface.beginPath();
		drawingSurface.arc(Math.floor(player.xPos), Math.floor(player.yPos), player.radius, 0, 2 * Math.PI, false);
		drawingSurface.fillStyle = player.color;
      	drawingSurface.fill();
	},

	spawn: function() {
		player.xPos = canvas.width / 2;
		player.yPos = canvas.height / 2;
	},

	/**
		Is the player in a corner, if so we shouldn't spawn drones there
	*/
	inCorner: function() {
		var marginOfError = 1.7;
		if (player.xPos < (canvas.width / 4) * marginOfError) {
			if (player.yPos < (canvas.height / 4) * marginOfError) {
				return 0; //top left
			} else if (player.yPos > (canvas.height * (3 / 4)) * (1 / marginOfError)) {
				return 3; //bottom left
			}
		} else if (player.xPos > (canvas.width * (3 / 4)) * (1 / marginOfError)) {
			if (player.yPos < (canvas.height / 4) * marginOfError) {
				return 1; //top right
			} else if (player.yPos > (canvas.height * (3 / 4)) * (1 / marginOfError)) {
				return 2; //bottom right
			}
		}
		return -1;
	}	
};

/***********************************************	GATE 		*******************/
var INITIAL_GATE_SPAWN_AMOUNT = 1;
var gateSpawnAmount = INITIAL_GATE_SPAWN_AMOUNT;
var gateSpawnTimer;
var gateNextSpawnTime;
var gate = {
	imgSrc: "gate.png",

	sourceX: 0,
	sourceY: 0,
	sourceWidth: 1000,
	sourceHeight: 150,

	width: 100,
	height: 15,
	xPos: 0,
	yPos: 0,
	blastRadius: 200,
	
	rotation: 0,
	rotationSpeed: 0.6, //degrees per frame

	movementSpeed: 6,
	xVel: 0,
	yVel: 0,

	updatePosition: function() {
		this.rotation += this.rotationSpeed;

		if (this.checkBorderBallCollision() && endable) {
			currentGameState = GAME_STATE.ENDED;
		} else if (this.checkCollisionWith(player)) {
      		this.explode();
      	}
	},

	/*endPts: function() { //draws the endpts of the gate (uncomment in draw() function below)
		var angle = this.rotation * Math.PI / 180;
		var halfLength = (this.width / 2) - 8; //the offset is to center the endPt on the gate ball
		
		endX = this.xPos + (halfLength)*(Math.cos(angle));
		endY = this.yPos + (halfLength)*(Math.sin(angle));
		endX1 = this.xPos - (halfLength)*(Math.cos(angle));
		endY1 = this.yPos - (halfLength)*(Math.sin(angle));
		
		var endRadius = 8;
		drawingSurface.beginPath();
		drawingSurface.arc(Math.floor(endX), Math.floor(endY), endRadius, 0, 2 * Math.PI, false);
		drawingSurface.fillStyle = "red";
      	drawingSurface.fill();

		drawingSurface.beginPath();
		drawingSurface.arc(Math.floor(endX1), Math.floor(endY1), endRadius, 0, 2 * Math.PI, false);
		drawingSurface.fillStyle = "green";
      	drawingSurface.fill();
	},*/

	draw: function() {
		//Save the current state of the drawing surface before it's rotated
		drawingSurface.save();
		//Rotate the canvas
		drawingSurface.translate( // this shifts the whole canvas to be zeroed at the center of the rotated object 
			Math.floor(this.xPos), 
			Math.floor(this.yPos) 
		);
		drawingSurface.rotate(this.rotation * Math.PI / 180);
		
		//Stamp the image of the rotated object (which, since the canvas is zeroed about it’s center, should be drawn half its height upwards and half its width leftwards)
		drawingSurface.drawImage(
			gateImage,
			this.sourceX, this.sourceY,
			this.sourceWidth, this.sourceHeight,
			Math.floor(-this.width / 2), Math.floor(-this.height / 2),
			this.width, this.height
		);

		//Restore the drawing surface to its state before it was rotated (but this time with the rotated sprite still in place)
		drawingSurface.restore();

		//this.endPts();
		/*
		//also draw a dot at the center of the gate)
		drawingSurface.beginPath();
		drawingSurface.arc(Math.floor(this.xPos), Math.floor(this.yPos), 5, 0, 2 * Math.PI, false);
		drawingSurface.fillStyle = "white";
      	drawingSurface.fill();
      	*/
	},

	/**
		Check the angle
	*/
	checkCollisionWith: function(circle) { 
		//console.log(Math.atan2(this.yPos, this.xPos));
		var dx = circle.xPos - this.xPos;
		var dy = circle.yPos - this.yPos;
		var distCircleGate = Math.sqrt((dx*dx) + (dy*dy));
		var angleCircleGate = Math.atan2(dy, dx) - (this.rotation * Math.PI / 180);
		if (Math.abs(distCircleGate * Math.cos(angleCircleGate)) < this.width / 2) {
			if (Math.abs(distCircleGate * Math.sin(angleCircleGate)) < circle.radius) {
				return true;
			}
		}
		return false;
	},

	checkBorderBallCollision: function(circle) {
		var angle = this.rotation * Math.PI / 180;
		var halfLength = (this.width / 2) - 8; //the offset is to center the endPt on the gate ball
		
		endX = this.xPos + (halfLength)*(Math.cos(angle));
		endY = this.yPos + (halfLength)*(Math.sin(angle));
		var ball = Object.create(gateEnd);
		ball.xPos = endX;
		ball.yPos = endY;

		endX1 = this.xPos - (halfLength)*(Math.cos(angle));
		endY1 = this.yPos - (halfLength)*(Math.sin(angle));
		var bomb = Object.create(gateEnd);
		bomb.xPos = endX1;
		bomb.yPos = endY1;

		if (collisionCircle(ball, player) || collisionCircle(bomb, player)) {
			return true;
		}
		return false;
	},

	explode: function() {
		/**
			Check for drones in a certain radius and kill them
		*/
		for (var i = drones.length - 1; i >= 0; i--) { //start from end, so you don't skip any
			var d = drones[i];
			var dx = d.xPos - this.xPos;
			var dy = d.yPos - this.yPos;

			var dist2 = dx*dx + dy*dy;

			if (dist2 < (this.blastRadius * this.blastRadius)) { //if within the blastRadius, kill the drone
				d.die();
			}
		}

		gate_explodeMP3.volume = 1;
		gate_explodeMP3.play();

		removeObjectFromArray(this, gates);
	},

	spawn: function(corner) { 
		var maxX = canvas.width - this.width;
		var maxY = canvas.height - this.width;
		var minX = this.width;
		var minY = this.width;

		var randX = Math.floor(Math.random() * (maxX - minX) + minX); 
		var randY = Math.floor(Math.random() * (maxY - minY) + minY); 
		var randAngle = Math.floor(Math.random() * 180);

		//Make sure gate doesn't spawn too close to player
		var dx = player.xPos - randX;
		var dy = player.yPos - randY;
		var distToPlayer = dx*dx + dy*dy;
		var minDist = player.radius + (this.width / 2) + 30; //spawn gate minimum of ____ pixels away from player
		while (distToPlayer < minDist*minDist) { //to circumvent sqrting
			randX = Math.floor(Math.random() * (maxX - minX) + minX); 
			randY = Math.floor(Math.random() * (maxY - minY) + minY); 
		}

		this.xPos = randX;
		this.yPos = randY;
		this.rotation = randAngle;
		gates.push(this); //this is an array in the game.html
	}
};
/**
	The border bombs at the ends of the gate; player collision with these is fatal.
*/
var gateEnd = { 
	xPos: 0,
	yPos: 0,
	radius: 7
}
function spawnGate() {
	for (var i = 0; i < gateSpawnAmount; i++) {
		var newGate = Object.create(gate);
		newGate.spawn();	
	}

	if (typeof gateSpawnTimer != 'undefined') {
		clearTimeout(gateSpawnTimer);	
	}

	var maxFreq = 6000 / gameSpeed;
	var minFreq = 2000 / gameSpeed;
	gateNextSpawnTime = Math.floor(Math.random() * (maxFreq - minFreq) + minFreq);
}

File: sprites.py
import arcade
import math
import numpy as np
import random
from game_functional import *

class Player:
    def __init__(self, game):
        self.x = 0
        self.y = 0
        self.dx = 0
        self.dy = 0
        self.target_x = None
        self.target_y = None

        self.game = game
        self.radius = 12
        self.color = arcade.color.WHITE
    
        self.movement_speed = 7

    def move_to(self, new_x, new_y): 
        """ Decomposes velocity into appropriate x and y based off sum total movement speed and sets target x and y """
        dx = new_x - self.x
        dy = new_y - self.y
        angle = math.atan2(dy, dx)

        self.target_x = new_x
        self.target_y = new_y
        self.dx = self.movement_speed * math.cos(angle)
        self.dy = self.movement_speed * math.sin(angle)

    def update(self):
        if self.target_x is not None:
            #stop the player if he's within a certain distance of the mouse
            error_factor = 0.32
            if (abs(self.x - self.target_x) < self.radius * error_factor):
                self.dx = 0
            if (abs(self.y - self.target_y) < self.radius * error_factor):
                self.dy = 0

        #Move the player and keep it inside screen boundaries
        self.x = max(0 + self.radius, min(self.x + self.dx, self.game.window_width - self.radius))
        self.y = max(0 + self.radius, min(self.y + self.dy, self.game.window_height - self.radius))

    def draw(self):
        """ Draws player as circle """
        arcade.draw_circle_filled(self.x, self.y, self.radius, self.color)

    def spawn(self):
        """ Spawns player in middle of screen """
        self.x = self.game.window_width / 2
        self.y = self.game.window_height / 2

    def corner(self):
        """ Returns which corner the player is in (we shouldn't spawn drones there) """
        margin_of_error = 1.7
        if (self.x < (self.game.window_width / 4) * margin_of_error):
            if (self.y < (self.game.window_height / 4) * margin_of_error):
                return self.game.TOP_LEFT_CORNER #top left
            elif (self.y > (self.game.window_height * (3 / 4)) * (1 / margin_of_error)):
                return self.game.BOTTOM_LEFT_CORNER #bottom left
        elif (self.x > (self.game.window_width * (3 / 4)) * (1 / margin_of_error)):
            if (self.y < (self.game.window_height / 4) * margin_of_error):
                return self.game.TOP_RIGHT_CORNER #top right
            elif (self.y > (self.game.window_height * (3 / 4)) * (1 / margin_of_error)):
                return self.game.BOTTOM_RIGHT_CORNER #bottom right
        return -1

class Drones:
    def __init__(self, window):
        self.game = window
        self.INITIAL_DRONE_SPAWN_AMOUNT = 20
        self.drone_spawn_amount = self.INITIAL_DRONE_SPAWN_AMOUNT
        self.drone_next_spawn_time = 0 #time till next spawn (random b/t 3-5 seconds)
        self.drones = [] #list of drones
        self.drone_render_list = arcade.ShapeElementList()
        self.update_delay = 4
        self.update_counter = 0

    def draw(self):
        """ Draws all the drones """
        self.drone_render_list.draw()

    def update(self):
        if self.update_counter % self.update_delay == 0:
            self.drone_render_list = arcade.ShapeElementList()
            for i in range(len(self.drones)):
                drone = self.drones[i]
                drone.update()   

                drone_shape = arcade.create_ellipse_filled(drone.x, drone.y, drone.radius, drone.radius, drone.color)
                self.drone_render_list.append(drone_shape)

                #also pairwise block overlap of the drones' positions (so they don't collide with each other)
                for j in range(i+1, len(self.drones)):
                    d1 = self.drones[j]
                    block_circle_overlap(drone, d1)

                #check for player collision with a drone
                if has_collided(self.game.player, drone) and not drone.dead and self.game.endable:
                    self.game.current_game_state = self.game.GAME_STATE['ENDED']

            self.drone_next_spawn_time -= math.ceil(1000 / 60) #assuming this is called at 60fps
            if self.drone_next_spawn_time <= 0:
                self.spawn_drones() #resets self.drone_next_spawn_time
        self.update_counter += 1

    def spawn_drones(self):
        """
        drone_spawnMP3.volume = 0.3
        drone_spawnMP3.play()
        """
        drone_spawn_loc = math.floor(random.random() * 4)
        if drone_spawn_loc == self.game.player.corner(): #don't spawn drones in the same corner as the player
            drone_spawn_loc = (drone_spawn_loc + 1) % 4

        for _ in range(self.drone_spawn_amount):
            new_drone = Drone(self.game)
            new_drone.spawn(drone_spawn_loc)  
            self.drones.append(new_drone)

        self.reset_next_spawn_time()

    def reset_next_spawn_time(self):
        #adjusted to self.game.game_speed
        max_msec = 1250 #5000
        min_msec = 750 #3500
        max_freq = max_msec / self.game.game_speed
        min_freq = min_msec / self.game.game_speed
        self.drone_next_spawn_time = math.floor(random.random() * (max_freq - min_freq) + min_freq) #reset next spawn time

class Drone:
    def __init__(self, window):
        self.x = 0
        self.y = 0
        self.dx = 0
        self.dy = 0
        self.target_x = None
        self.target_y = None

        self.game = window
        self.radius = 9
        self.color = arcade.color.CYAN
    
        self.movement_speed = 20
        self.dead = False #to prevent a weird afterdeath collision bug
        self.spawn_cooldown = 175 #700 #give player time to react to drones spawning
        self.pts = 15 #how many pts the player gets for killing this drone

    def move_to(self, new_x, new_y): 
        #decomposes velocity into appropriate x and y based off drones sum total movement speed and sets target x and y
        dx = new_x - self.x
        dy = new_y - self.y
        angle = math.atan2(dy, dx)

        self.dx = self.movement_speed * math.cos(angle)
        self.dy = self.movement_speed * math.sin(angle)

    def update(self):
        if self.spawn_cooldown > 0:
            self.color = arcade.color.BLUE #"#33A0FF"
            self.spawn_cooldown -= math.ceil(1000/60) #1 sec spawn time
        else:
            self.color = arcade.color.CYAN
            self.move_to(self.game.player.x, self.game.player.y) #target player

            #Move the drone and keep it inside screen boundaries
            self.x = max(0 + self.radius, min(self.x + self.dx, self.game.window_width - self.radius))
            self.y = max(0 + self.radius, min(self.y + self.dy, self.game.window_width - self.radius))

    def draw(self):
        """ Draws drone as circle """
        arcade.draw_circle_filled(self.x, self.y, self.radius, self.color)

    def spawn(self, corner):
        min_x = 0
        max_x = 0
        min_y = 0
        max_y = 0
        #spawn in one of the four corners [TOP_LEFT_CORNER, TOP_RIGHT_CORNER, BOTTOM_RIGHT_CORNER, BOTTOM_LEFT_CORNER]
        if corner == self.game.TOP_LEFT_CORNER:
            min_x = 0
            max_x = self.game.window_width / 4
            min_y = 0
            max_y = self.game.window_height / 4
        elif corner == self.game.TOP_RIGHT_CORNER:
            min_x = self.game.window_width * (3 / 4)
            max_x = self.game.window_width
            min_y = 0
            max_y = self.game.window_height / 4
        elif corner == self.game.BOTTOM_RIGHT_CORNER: 
            min_x = self.game.window_width * (3 / 4)
            max_x = self.game.window_width
            min_y = self.game.window_height * (3 / 4)
            max_y = self.game.window_height
        elif corner == self.game.BOTTOM_LEFT_CORNER: 
            min_x = 0
            max_x = self.game.window_width / 4
            min_y = self.game.window_height * (3 / 4)
            max_y = self.game.window_height

        #offset by radius so that entire drone appears within boundaries (not half the radius outside)
        max_x -= self.radius
        max_y -= self.radius
        min_x += self.radius
        min_y += self.radius

        rand_x = math.floor(random.random() * (max_x - min_x) + min_x) #randomX in range [drone.radius, max_x]
        rand_y = math.floor(random.random() * (max_y - min_y) + min_y) 

        self.x = rand_x
        self.y = rand_y

    def die(self, drones):
        self.dead = true
        drones.remove(self)

        """
        newExplosion = Object.create(explosion)
        newExplosion.x = self.x - self.radius
        newExplosion.y = self.y - self.radius
        #newExplosion.display_size = self.radius
        explosions.push(newExplosion)
        """

        self.game.game_score += self.pts

class Explosions:

class Explosion:
    def __init__(self,):
        self.img_size = 92
        self.display_size = 20 #pixels
        self.COLUMNS = 3 #= # frames per row on the tile sheet
        self.number_of_frames = 17 #18 - 1
        self.current_frame = 0
        self.source_x = 0
        self.source_y = 0

        self.x = 0
        self.y = 0

    def update(self):
        #Find the frame's correct column and row on the tilesheet
        self.source_x = math.floor(self.current_frame % self.COLUMNS) * self.img_size
        self.source_y = math.floor(self.current_frame / self.COLUMNS) * self.img_size

        if self.current_frame < self.number_of_frames:
            self.current_frame++
        else:
            removeObjectFromArray(self, explosions)

    def draw():
        #Draw the monster's current animation frame
        drawingSurface.drawImage (
            burstImage,
            self.source_x, self.source_y, self.img_size, self.img_size,
            self.xPos, self.yPos, self.display_size, self.display_size
        )
}